RFC-001-BAM.md

```diff
@@ -119,15 +125,64 @@ The messages are also not dynamic as their structure needs to be known at compil
+
+The libp2p stack of COMIT is configured as follows:
+
+- tcp as the transport layer
+- yamux as the multiplexing module
+- secio as the encryption module
+
+## Multiplexing
+
+Communication in a peer-to-peer network typically requires some form of multiplexing to allow every party to act as the initiator of a session if they want to actively send messages to another party.
+This section first briefly explains, how such multiplexing can be achieved if all you have are plain network sockets.
+As a second step, we will quickly cover, how the `substream` mechanism of libp2p affects this and thereby simplifies our application protocol.
+
+### Multiplexing with plain sockets
+
+With traditional network sockets, there is always a listener and a dialer.
+The listening party has to have a port open the dialling party can connect to.
+If one party is behind a NAT, the only way to establish a network connection is for this party to dial to the other one.
+Establishing such a connection is expensive.
+We have to wait for the TCP handshake to complete and if there is transport encryption in place, keys might need to be exchanged before the first actual payload can be sent to the other party.
+Due to this cost and network limitations like NAT, it is good to reuse a connection between two parties once we have established one.
+This is where multiplexing comes into play.
+If both parties want to be able to send and receive messages at any point in time, those messages need to carry metadata so that the receiving end can correctly identify and associate messages.
+
+If all you have are plain network sockets, such a multiplexing strategy needs to be backed into your application protocol, which is a none trivial increase in complexity.
+In fact, a former version of the COMIT messaging protocol (BAM!) had support for such functionality because we were not yet building on top of libp2p back then.
+
+### Multiplexing with libp2p substreams
+
+Libp2p has the concept of substreams.
+A substream is a virtual socket for a specific protocol between two nodes.
+Opening a new substream is cheap compared to an actual network connection.
+It is also independent of the underlying physical network topology, i.e. it does not matter which party opened the actual physical network connection:
+Every party can open a new substream at any point in time.
+
+Given that, we define the following rules for sending and receiving messages in the COMIT network:
+
+1. To actively send a message, a node MUST open a new substream.
+2. The node opening the substream MUST send the first message.
+3. The first message MUST identify the kind of messages being sent on this substream.
+More on this in section [messages](#messages).
+
+The multiplexing solution of substreams offered by libp2p greatly reduces the complexity of our application protocol because we no longer have to explicitly associate messages with each other but instead can rely on the ordering of the messages within a substream. 
```

Comment:
```suggestion
The multiplexing solution of substreams offered by libp2p greatly reduces the complexity of our application protocol because we no longer have to explicitly associate messages with each other but instead can rely on the ordering of the messages within a substream.
```